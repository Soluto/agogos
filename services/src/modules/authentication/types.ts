import { ClientOptions } from 'jwks-rsa';

export interface IssuerConfig {
  authority: string;
  jwksConfig?: Partial<ClientOptions>;
  audience?: string | string[];
  authenticatedPaths: string[];
  description?: string;
}

export interface AuthenticationConfig {
  jwt?: Record<string, IssuerConfig>;
  anonymous?: {
    publicPaths: string[];
    rejectAuthorizationHeader?: boolean;
  };
}
