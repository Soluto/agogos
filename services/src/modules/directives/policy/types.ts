import {PolicyArgsObject, PolicyAttachments} from '../../resource-repository/types';

export type Policy = {
    namespace: string;
    name: string;
    args?: PolicyArgsObject;
};

// args here contain the final values after param injection
export type PolicyExecutionContext = {
    namespace: string;
    name: string;
    policyAttachments: PolicyAttachments;
    args?: PolicyArgsObject;
    query?: QueryResults;
};

export type QueryResults = {
    [name: string]: any;
};

export type PolicyExecutionResult = {
    done: boolean;
    allow?: boolean;
    query?: {
        type: string;
        code: string;
    };
};

export type GraphQLArguments = {
    [name: string]: any;
};
