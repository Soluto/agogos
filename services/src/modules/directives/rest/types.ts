export interface KeyValue {
  key: string;
  value: string;
}

export interface RestParams {
  url: string;
  method?: string;
  body?: string;
  bodyArg?: string;
  query?: KeyValue[];
  headers?: KeyValue[];
  timeoutMs?: number;
}
