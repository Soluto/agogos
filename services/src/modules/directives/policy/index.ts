import { PolicyDefinition } from '../../resource-repository';
import { sdl as policySdl, PolicyDirective } from './policy';
import { sdl as policiesSdl, PoliciesDirective, resolvers as policyScalarResolvers } from './policies';
import { sdl as policyQuerySdl, PolicyQueryDirective } from './policy-query';
import PolicyExecutor from './policy-executor';
import { Policy, LoadedPolicy } from './types';

export {
  PolicyExecutor,
  policySdl,
  policiesSdl,
  policyQuerySdl,
  PolicyDirective,
  PoliciesDirective,
  PolicyQueryDirective,
  policyScalarResolvers,
};

export type AuthorizationConfig = {
  policies: PolicyDefinition[];
  policyAttachments?: Record<string, LoadedPolicy>;
  policyExecutor: PolicyExecutor;
  basePolicy?: Policy;
};
