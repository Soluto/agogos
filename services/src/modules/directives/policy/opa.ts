// @ts-ignore opa-wasm already has TS typings merged, but not yet published on npm
import * as Rego from '@open-policy-agent/opa-wasm';
import { getCompiledFilename } from '../../opa-helper';
import { PolicyExecutionContext, PolicyExecutionResult, QueryResults } from './types';
import { PolicyArgsObject } from '../../resource-repository';

export async function evaluate(ctx: PolicyExecutionContext): Promise<PolicyExecutionResult> {
  const policy = await getWasmPolicy(ctx);
  const input = getInput(ctx);

  const result = policy.evaluate(input)?.[0]?.result;

  return { done: true, allow: result?.allow };
}

async function getWasmPolicy(ctx: PolicyExecutionContext): Promise<any> {
  const filename = getCompiledFilename({ namespace: ctx.namespace, name: ctx.name });
  const wasm = ctx.policyAttachments[filename];

  const rego = new Rego();
  return rego.load_policy(wasm);
}

function getInput(ctx: PolicyExecutionContext): PolicyOpaInput {
  const input: PolicyOpaInput = {};

  if (ctx.args) input.args = ctx.args;
  if (ctx.query) input.query = ctx.query;

  return input;
}

type PolicyOpaInput = {
  args?: PolicyArgsObject;
  query?: QueryResults;
};
