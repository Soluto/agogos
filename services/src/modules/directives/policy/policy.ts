import { GraphQLResolveInfo, GraphQLField, defaultFieldResolver, GraphQLObjectType, FieldNode } from 'graphql';
import { SchemaDirectiveVisitor } from 'graphql-tools';
import { gql } from 'apollo-server-core';
import { RequestContext } from '../../context';
import { GraphQLArguments, Policy } from './types';

export class PolicyDirective extends SchemaDirectiveVisitor {
  visitObject(object: GraphQLObjectType<unknown, RequestContext>) {
    const policy = this.args as Policy;
    const originalResolveObject = object.resolveObject;

    object.resolveObject = async (
      source: unknown,
      fields: Record<string, ReadonlyArray<FieldNode>>,
      context: RequestContext,
      info: GraphQLResolveInfo
    ) => {
      if (!context.ignorePolicies) {
        await context.authorizationConfig.policyExecutor.validatePolicy(policy, source, {}, context, info);
      }

      return originalResolveObject ? await originalResolveObject(source, fields, context, info) : source;
    };
  }

  visitFieldDefinition(field: GraphQLField<unknown, RequestContext>) {
    const policy = this.args as Policy;
    const originalResolve = field.resolve || defaultFieldResolver;

    field.resolve = async (
      source: unknown,
      args: GraphQLArguments,
      context: RequestContext,
      info: GraphQLResolveInfo
    ) => {
      if (!context.ignorePolicies) {
        await context.authorizationConfig.policyExecutor.validatePolicy(policy, source, args, context, info);
      }

      return originalResolve.call(field, source, args, context, info);
    };
  }
}

export const sdl = gql`
  directive @policy(namespace: String!, name: String!, args: JSONObject) on OBJECT | FIELD_DEFINITION
`;
