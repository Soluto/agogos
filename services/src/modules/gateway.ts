import * as fastify from 'fastify';
import { ApolloServer, Config } from 'apollo-server-fastify';
import { Observable } from 'rxjs';
import { createGraphQLService } from './graphql-service';
import { RESTDirectiveDataSource } from './directives/rest';
import { ResourceGroup } from './resource-repository';
import { ExportTrackingExtension } from './exports';
import {
  createBasicPolicyPlugin,
  createLoggingPlugin,
  createMetricsPlugin,
  createArgumentInjectionPlugin,
} from './plugins';

export interface GatewayConfig extends Config {
  resourceGroups: Observable<ResourceGroup>;
  fastifyInstance: Pick<fastify.FastifyInstance, 'metrics'>;
}

export function createStitchGateway(config: GatewayConfig) {
  const { resourceGroups, fastifyInstance, ...apolloConfig } = config;
  const gateway = createGraphQLService({ resourceGroups });
  const server = new ApolloServer({
    ...apolloConfig,
    gateway,
    extensions: [() => new ExportTrackingExtension()],
    plugins: [
      createBasicPolicyPlugin,
      createLoggingPlugin(),
      createMetricsPlugin(fastifyInstance),
      createArgumentInjectionPlugin(),
    ],
    subscriptions: false,
    context(request: fastify.FastifyRequest) {
      const ctx = {
        request,
      };
      return ctx;
    },
    dataSources() {
      return {
        rest: new RESTDirectiveDataSource(),
      };
    },
  });

  return {
    server,
    dispose: async () => {
      await server.stop();
      gateway.dispose();
    },
  };
}
