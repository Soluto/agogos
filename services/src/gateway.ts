import * as fastify from 'fastify';
import * as fastifyMetrics from 'fastify-metrics';

import * as config from './modules/config';
import logger from './modules/logger';
import {handleSignals, handleUncaughtErrors} from './modules/shutdownHandler';
import {createStitchGateway} from './modules/gateway';
import {pollForUpdates, S3ResourceRepository} from './modules/resource-repository';

async function run() {
    logger.info('Stitch gateway booting up...');

    const resourceRepository = new S3ResourceRepository();

    const {server, dispose} = createStitchGateway({
        resourceGroups: pollForUpdates(resourceRepository, config.resourceUpdateInterval),
        tracing: config.enableGraphQLTracing,
        playground: config.enableGraphQLPlayground,
        introspection: config.enableGraphQLIntrospection,
    });

    const app = fastify();
    app.register(fastifyMetrics, {endpoint: '/metrics'});
    app.register(server.createHandler({path: '/graphql'}));
    const address = await app.listen(config.httpPort, '0.0.0.0');
    logger.info({address}, 'Stitch gateway started');

    handleSignals(dispose);
    handleUncaughtErrors();
}

// Only run when file is being executed, not when being imported
if (require.main === module) {
    run();
}
