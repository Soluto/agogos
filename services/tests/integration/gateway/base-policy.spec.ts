import { ApolloServerTestClient, createTestClient } from 'apollo-server-testing';
import { print } from 'graphql';
import { gql } from 'apollo-server-core';
import * as Rx from 'rxjs';
import GraphQLErrorSerializer from '../../utils/graphql-error-serializer';
import { createStitchGateway } from '../../../src/modules/gateway';
import { ResourceGroup, PolicyDefinition, PolicyType } from '../../../src/modules/resource-repository';
import { beforeEachDispose } from '../before-each-dispose';
import { LoadedPolicy } from '../../../src/modules/directives/policy/opa';
import { Policy } from '../../../src/modules/directives/policy/types';

const createPolicyAttachment = (allow: boolean): LoadedPolicy => ({ evaluate: () => [{ result: { allow } }] });

const policies: PolicyDefinition[] = [
  {
    metadata: {
      namespace: 'internal',
      name: 'base_policy',
    },
    type: PolicyType.opa,
    code: `Rego code`,
  },
  {
    metadata: {
      namespace: 'ns',
      name: 'field_policy',
    },
    type: PolicyType.opa,
    code: `Rego code`,
  },
  {
    metadata: {
      namespace: 'ns',
      name: 'override_base_policy',
    },
    shouldOverrideBasePolicy: true,
    type: PolicyType.opa,
    code: `Rego code`,
  },
];

const basePolicy: Policy = {
  namespace: 'internal',
  name: 'base_policy',
};

const testCases: [string, ResourceGroup & { etag: string }][] = [
  [
    'Deny on base policy, allow on field definition',
    {
      etag: 'etag1',
      basePolicy,
      policies,
      policyAttachments: {
        ['internal-base_policy.wasm']: createPolicyAttachment(false),
        ['ns-field_policy.wasm']: createPolicyAttachment(true),
      },
      upstreams: [],
      upstreamClientCredentials: [],
      schemas: [
        {
          metadata: {
            namespace: 'ns',
            name: 'main',
          },
          schema: print(gql`
            type Query {
              foo: String @stub(value: "FOO") @policy(namespace: "ns", name: "field_policy")
              bar: String @stub(value: "BAR")
            }
          `),
        },
      ],
    },
  ],
  [
    'Allow on base policy, deny on field definition',
    {
      etag: 'etag2',
      basePolicy,
      policies,
      policyAttachments: {
        ['internal-base_policy.wasm']: createPolicyAttachment(true),
        ['ns-field_policy.wasm']: createPolicyAttachment(false),
      },
      upstreams: [],
      upstreamClientCredentials: [],
      schemas: [
        {
          metadata: {
            namespace: 'ns',
            name: 'main',
          },
          schema: print(gql`
            type Query {
              foo: String @stub(value: "FOO") @policy(namespace: "ns", name: "field_policy")
              bar: String @stub(value: "BAR")
            }
          `),
        },
      ],
    },
  ],
  [
    'Allow on base policy, allow on field definition',
    {
      etag: 'etag3',
      basePolicy,
      policies,
      policyAttachments: {
        ['internal-base_policy.wasm']: createPolicyAttachment(true),
        ['ns-override_base_policy.wasm']: createPolicyAttachment(true),
      },
      upstreams: [],
      upstreamClientCredentials: [],
      schemas: [
        {
          metadata: {
            namespace: 'ns',
            name: 'main',
          },
          schema: print(gql`
            type Query {
              foo: String @stub(value: "FOO") @policy(namespace: "ns", name: "override_base_policy")
              bar: String @stub(value: "BAR")
            }
          `),
        },
      ],
    },
  ],
  [
    'Deny on base policy, allow on field definition with overrideBasePolicy',
    {
      etag: 'etag4',
      basePolicy,
      policies,
      policyAttachments: {
        ['internal-base_policy.wasm']: createPolicyAttachment(false),
        ['ns-override_base_policy.wasm']: createPolicyAttachment(true),
      },
      upstreams: [],
      upstreamClientCredentials: [],
      schemas: [
        {
          metadata: {
            namespace: 'ns',
            name: 'main',
          },
          schema: print(gql`
            type Query {
              foo: String @stub(value: "FOO") @policy(namespace: "ns", name: "override_base_policy")
              bar: String @stub(value: "BAR")
            }
          `),
        },
      ],
    },
  ],
];

describe.each(testCases)('Base Policy Tests', (testName, resourceGroup) => {
  let client: ApolloServerTestClient;

  const query = gql`
    query {
      foo
      bar
    }
  `;

  beforeEachDispose(() => {
    const stitch = createStitchGateway({ resourceGroups: Rx.of(resourceGroup) });
    client = createTestClient(stitch.server);

    return () => {
      return stitch.dispose();
    };
  });

  beforeAll(() => {
    expect.addSnapshotSerializer(GraphQLErrorSerializer);
  });

  test(testName, async () => {
    const response = await client.query({ query });
    expect(response).toMatchSnapshot();
  });
});
