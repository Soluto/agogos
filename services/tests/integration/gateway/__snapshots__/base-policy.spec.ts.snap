// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base Policy Tests Allow on base policy, allow on field definition 1`] = `
Object {
  "data": Object {
    "bar": "BAR",
    "foo": "FOO",
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Base Policy Tests Allow on base policy, deny on field definition 1`] = `
{
  "errors": [
    {
      "name": "GraphQLError",
      "message": "Unauthorized by policy field_policy in namespace ns",
      "errorCode": "INTERNAL_SERVER_ERROR",
      "path": [
        "foo"
      ]
    }
  ],
  "data": {
    "foo": null,
    "bar": "BAR"
  }
}
`;

exports[`Base Policy Tests Deny on base policy, allow on field definition 1`] = `
{
  "errors": [
    {
      "name": "GraphQLError",
      "message": "Unauthorized by policy base_policy in namespace internal",
      "errorCode": "INTERNAL_SERVER_ERROR",
      "path": [
        "foo"
      ]
    },
    {
      "name": "GraphQLError",
      "message": "Unauthorized by policy base_policy in namespace internal",
      "errorCode": "INTERNAL_SERVER_ERROR",
      "path": [
        "bar"
      ]
    }
  ],
  "data": {
    "foo": null,
    "bar": null
  }
}
`;

exports[`Base Policy Tests Deny on base policy, allow on field definition with overrideBasePolicy 1`] = `
{
  "errors": [
    {
      "name": "GraphQLError",
      "message": "Unauthorized by policy base_policy in namespace internal",
      "errorCode": "INTERNAL_SERVER_ERROR",
      "path": [
        "bar"
      ]
    }
  ],
  "data": {
    "foo": "FOO",
    "bar": null
  }
}
`;
